cpp/src/parquet/page_index.h

```diff
@@ -0,0 +1,143 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <cstdint>
+#include <map>
+#include <memory>
+#include <set>
+#include <string>
+#include <vector>
+
+#include "parquet/platform.h"
+#include "parquet/statistics.h"
+#include "parquet/types.h"
+
+namespace parquet {
+
+/// \brief BoundaryOrder is a proxy around format::BoundaryOrder.
+enum class PARQUET_EXPORT BoundaryOrder { UNORDERED = 0, ASCENDING = 1, DESCENDING = 2 };
+
+/// \brief ColumnIndex is a proxy around format::ColumnIndex.
+class PARQUET_EXPORT ColumnIndex {
+ public:
+  /// \brief Create a ColumnIndex from a serialized thrift message.
+  static std::unique_ptr<ColumnIndex> Make(const ColumnDescriptor* descr,
+                                           const void* serialized_index,
+                                           uint32_t* inout_index_len,
+                                           const ReaderProperties& properties);
+
+  virtual ~ColumnIndex() = default;
+
+  /// \brief Returns number of pages in this column index.
+  virtual int64_t num_pages() const = 0;
+
+  /// \brief Returns if all values are null in a single page.
+  virtual bool null_page(int64_t page_id) const = 0;
+
+  /// \brief Returns whether both min_values and max_values are
+  /// orderd and if so, in which direction.
+  virtual BoundaryOrder boundary_order() const = 0;
+
+  /// \brief Returns if null count is available.
+  virtual bool HasNullCount() const = 0;
```

Comment:
We don't have a very well-defined policy on casing, but generally simple accessors should be lowercase.
```suggestion
  virtual bool has_null_counts() const = 0;
```